File: backend/app/forms/expr.py
"""Tiny expression language for computed form fields.

Mirrors frontend/src/forms/expr.ts; both are checked against
app/data/forms/expr_vectors.json.
"""
from __future__ import annotations

import re
from datetime import date
from typing import Any

class ExprError(Exception):
    pass

_TOKEN = re.compile(r"\s*(?:(\d+\.\d+|\d+)|('(?:[^'\\]|\\.)*')|(==|!=|<=|>=|[-+*/()<>,\[\]])|([A-Za-z_][A-Za-z0-9_]*))")

def _tokenize(src: str) -> list[tuple[str, Any]]:
    out, pos = [], 0
    while pos < len(src):
        m = _TOKEN.match(src, pos)
        if not m or m.end() == pos:
            if src[pos:].strip() == "":
                break
            raise ExprError(f"bad token at {pos}: {src[pos:pos+10]!r}")
        pos = m.end()
        num, s, op, ident = m.groups()
        if num is not None:
            out.append(("num", float(num) if "." in num else int(num)))
        elif s is not None:
            out.append(("str", s[1:-1].replace("\\'", "'")))
        elif op is not None:
            out.append(("op", op))
        else:
            out.append(("id", ident))
    out.append(("eof", None))
    return out

class _Parser:
    def __init__(self, tokens):
        self.t, self.i = tokens, 0
    def peek(self):
        return self.t[self.i]
    def take(self, kind=None, val=None):
        k, v = self.t[self.i]
        if (kind and k != kind) or (val is not None and v != val):
            raise ExprError(f"expected {val or kind}, got {v!r}")
        self.i += 1
        return v
    # precedence climbing
    def parse(self):
        node = self.p_or()
        if self.peek()[0] != "eof":
            raise ExprError(f"unexpected {self.peek()[1]!r}")
        return node
    def p_or(self):
        n = self.p_and()
        while self.peek() == ("id", "or"):
            self.take(); n = ("or", n, self.p_and())
        return n
    def p_and(self):
        n = self.p_not()
        while self.peek() == ("id", "and"):
            self.take(); n = ("and", n, self.p_not())
        return n
    def p_not(self):
        if self.peek() == ("id", "not"):
            self.take(); return ("not", self.p_not())
        return self.p_cmp()
    def p_cmp(self):
        n = self.p_add()
        while self.peek()[0] == "op" and self.peek()[1] in ("==", "!=", "<", "<=", ">", ">="):
            op = self.take(); n = ("cmp", op, n, self.p_add())
        return n
    def p_add(self):
        n = self.p_mul()
        while self.peek()[0] == "op" and self.peek()[1] in ("+", "-"):
            op = self.take(); n = ("bin", op, n, self.p_mul())
        return n
    def p_mul(self):
        n = self.p_unary()
        while self.peek()[0] == "op" and self.peek()[1] in ("*", "/"):
            op = self.take(); n = ("bin", op, n, self.p_unary())
        return n
    def p_unary(self):
        if self.peek() == ("op", "-"):
            self.take(); return ("neg", self.p_unary())
        return self.p_primary()
    def p_primary(self):
        k, v = self.peek()
        if k == "num": self.take(); return ("lit", v)
        if k == "str": self.take(); return ("lit", v)
        if k == "op" and v == "(":
            self.take(); n = self.p_or(); self.take("op", ")"); return n
        if k == "op" and v == "[":
            self.take(); items = []
            while self.peek() != ("op", "]"):
                items.append(self.p_or())
                if self.peek() == ("op", ","): self.take()
            self.take("op", "]"); return ("list", items)
        if k == "id":
            self.take()
            if v in ("true", "false"): return ("lit", v == "true")
            if v == "null": return ("lit", None)
            if self.peek() == ("op", "("):
                self.take(); args = []
                while self.peek() != ("op", ")"):
                    args.append(self.p_or())
                    if self.peek() == ("op", ","): self.take()
                self.take("op", ")"); return ("call", v, args)
            return ("var", v)
        raise ExprError(f"unexpected {v!r}")

def _num(x):
    return 0 if x is None or x is False else (1 if x is True else x)

def _eval(node, scope):
    kind = node[0]
    if kind == "lit": return node[1]
    if kind == "var":
        if node[1] not in scope: raise ExprError(f"unknown identifier {node[1]}")
        return scope[node[1]]
    if kind == "list": return [_eval(n, scope) for n in node[1]]
    if kind == "neg": return -_num(_eval(node[1], scope))
    if kind == "not": return not _eval(node[1], scope)
    if kind == "and": return bool(_eval(node[1], scope)) and bool(_eval(node[2], scope))
    if kind == "or": return bool(_eval(node[1], scope)) or bool(_eval(node[2], scope))
    if kind == "cmp":
        op, a, b = node[1], _eval(node[2], scope), _eval(node[3], scope)
        if op == "==": return a == b
        if op == "!=": return a != b
        a, b = _num(a), _num(b)
        return {"<": a < b, "<=": a <= b, ">": a > b, ">=": a >= b}[op]
    if kind == "bin":
        op, a, b = node[1], _num(_eval(node[2], scope)), _num(_eval(node[3], scope))
        if op == "+": return a + b
        if op == "-": return a - b
        if op == "*": return a * b
        return 0 if b == 0 else a / b
    if kind == "call":
        name, args = node[1], node[2]
        if name in ("count", "sum"):
            if len(args) != 1: raise ExprError(f"{name} takes one argument")
            rows = scope.get("_rows") or []
            vals = [_eval(args[0], {**scope, **row}) for row in rows]
            return sum(1 for v in vals if v) if name == "count" else sum(_num(v) for v in vals)
        vals = [_eval(a, scope) for a in args]
        if name == "band":
            x, *bands, default = vals
            x = _num(x)
            for limit, label in bands:
                if x <= _num(limit): return label
            return default
        if name == "today": return date.today().isoformat()
        if name == "days_between":
            a, b = date.fromisoformat(str(vals[0])[:10]), date.fromisoformat(str(vals[1])[:10])
            return (b - a).days
        raise ExprError(f"unknown function {name}")
    raise ExprError(f"bad node {kind}")

_CACHE: dict[str, tuple] = {}

def evaluate(expr: str, scope: dict) -> Any:
    ast = _CACHE.get(expr)
    if ast is None:
        ast = _Parser(_tokenize(expr)).parse()
        _CACHE[expr] = ast
    return _eval(ast, scope)

File: frontend/src/forms/expr.ts
export class ExprError extends Error {}

type Tok = { k: 'num' | 'str' | 'op' | 'id' | 'eof'; v: string | number | null };
const RE = /\s*(?:(\d+\.\d+|\d+)|('(?:[^'\\]|\\.)*')|(==|!=|<=|>=|[-+*/()<>,\[\]])|([A-Za-z_][A-Za-z0-9_]*))/y;

function tokenize(src: string): Tok[] {
  const out: Tok[] = []; let pos = 0;
  while (pos < src.length) {
    RE.lastIndex = pos; const m = RE.exec(src);
    if (!m || m[0].length === 0) { if (src.slice(pos).trim() === '') break; throw new ExprError(`bad token at ${pos}`); }
    pos = RE.lastIndex;
    if (m[1] !== undefined) out.push({ k: 'num', v: Number(m[1]) });
    else if (m[2] !== undefined) out.push({ k: 'str', v: m[2].slice(1, -1).replace(/\\'/g, "'") });
    else if (m[3] !== undefined) out.push({ k: 'op', v: m[3] });
    else out.push({ k: 'id', v: m[4] });
  }
  out.push({ k: 'eof', v: null });
  return out;
}

type Node = ['lit', unknown] | ['var', string] | ['list', Node[]] | ['neg', Node] | ['not', Node]
  | ['and', Node, Node] | ['or', Node, Node] | ['cmp', string, Node, Node] | ['bin', string, Node, Node] | ['call', string, Node[]];

class Parser {
  i = 0;
  constructor(private t: Tok[]) {}
  peek() { return this.t[this.i]; }
  is(k: Tok['k'], v?: string) { const p = this.peek(); return p.k === k && (v === undefined || p.v === v); }
  take(k?: Tok['k'], v?: string) { const p = this.peek(); if ((k && p.k !== k) || (v !== undefined && p.v !== v)) throw new ExprError(`expected ${v ?? k}, got ${p.v}`); this.i++; return p.v; }
  parse(): Node { const n = this.or(); if (!this.is('eof')) throw new ExprError(`unexpected ${this.peek().v}`); return n; }
  or(): Node { let n = this.and(); while (this.is('id', 'or')) { this.take(); n = ['or', n, this.and()]; } return n; }
  and(): Node { let n = this.not(); while (this.is('id', 'and')) { this.take(); n = ['and', n, this.not()]; } return n; }
  not(): Node { if (this.is('id', 'not')) { this.take(); return ['not', this.not()]; } return this.cmp(); }
  cmp(): Node { let n = this.add(); while (this.is('op') && ['==', '!=', '<', '<=', '>', '>='].includes(String(this.peek().v))) { const op = String(this.take()); n = ['cmp', op, n, this.add()]; } return n; }
  add(): Node { let n = this.mul(); while (this.is('op', '+') || this.is('op', '-')) { const op = String(this.take()); n = ['bin', op, n, this.mul()]; } return n; }
  mul(): Node { let n = this.unary(); while (this.is('op', '*') || this.is('op', '/')) { const op = String(this.take()); n = ['bin', op, n, this.unary()]; } return n; }
  unary(): Node { if (this.is('op', '-')) { this.take(); return ['neg', this.unary()]; } return this.primary(); }
  primary(): Node {
    const p = this.peek();
    if (p.k === 'num' || p.k === 'str') { this.take(); return ['lit', p.v]; }
    if (p.k === 'op' && p.v === '(') { this.take(); const n = this.or(); this.take('op', ')'); return n; }
    if (p.k === 'op' && p.v === '[') { this.take(); const items: Node[] = []; while (!this.is('op', ']')) { items.push(this.or()); if (this.is('op', ',')) this.take(); } this.take('op', ']'); return ['list', items]; }
    if (p.k === 'id') {
      const name = String(this.take());
      if (name === 'true' || name === 'false') return ['lit', name === 'true'];
      if (name === 'null') return ['lit', null];
      if (this.is('op', '(')) { this.take(); const args: Node[] = []; while (!this.is('op', ')')) { args.push(this.or()); if (this.is('op', ',')) this.take(); } this.take('op', ')'); return ['call', name, args]; }
      return ['var', name];
    }
    throw new ExprError(`unexpected ${p.v}`);
  }
}

const num = (x: unknown): number => (x === null || x === undefined || x === false || x === '' ? 0 : x === true ? 1 : Number(x));

function ev(n: Node, scope: Record<string, unknown>): unknown {
  switch (n[0]) {
    case 'lit': return n[1];
    case 'var': if (!(n[1] in scope)) throw new ExprError(`unknown identifier ${n[1]}`); return scope[n[1]];
    case 'list': return n[1].map((x) => ev(x, scope));
    case 'neg': return -num(ev(n[1], scope));
    case 'not': return !ev(n[1], scope);
    case 'and': return Boolean(ev(n[1], scope)) && Boolean(ev(n[2], scope));
    case 'or': return Boolean(ev(n[1], scope)) || Boolean(ev(n[2], scope));
    case 'cmp': {
      const a = ev(n[2], scope), b = ev(n[3], scope);
      if (n[1] === '==') return (a ?? null) === (b ?? null);
      if (n[1] === '!=') return (a ?? null) !== (b ?? null);
      if (typeof a === 'string' && typeof b === 'string') {
        return n[1] === '<' ? a < b : n[1] === '<=' ? a <= b : n[1] === '>' ? a > b : a >= b;
      }
      const x = num(a), y = num(b);
      return n[1] === '<' ? x < y : n[1] === '<=' ? x <= y : n[1] === '>' ? x > y : x >= y;
    }
    case 'bin': {
      const a = num(ev(n[2], scope)), b = num(ev(n[3], scope));
      return n[1] === '+' ? a + b : n[1] === '-' ? a - b : n[1] === '*' ? a * b : b === 0 ? 0 : a / b;
    }
    case 'call': {
      const [, name, args] = n;
      if (name === 'count' || name === 'sum') {
        if (args.length !== 1) throw new ExprError(`${name} takes one argument`);
        const rows = (scope._rows as Record<string, unknown>[] | undefined) ?? [];
        const vals = rows.map((r) => ev(args[0], { ...scope, ...r }));
        return name === 'count' ? vals.filter(Boolean).length : vals.reduce<number>((s, v) => s + num(v), 0);
      }
      const vals = args.map((a) => ev(a, scope));
      if (name === 'band') {
        const x = num(vals[0]); const def = vals[vals.length - 1];
        for (const band of vals.slice(1, -1) as [unknown, unknown][]) if (x <= num(band[0])) return band[1];
        return def;
      }
      if (name === 'today') {
        const d = new Date();
        const yyyy = d.getFullYear();
        const mm = String(d.getMonth() + 1).padStart(2, '0');
        const dd = String(d.getDate()).padStart(2, '0');
        return `${yyyy}-${mm}-${dd}`;
      }
      if (name === 'days_between') {
        const a = new Date(String(vals[0]).slice(0, 10)), b = new Date(String(vals[1]).slice(0, 10));
        return Math.round((b.getTime() - a.getTime()) / 86400000);
      }
      throw new ExprError(`unknown function ${name}`);
    }
  }
}

const cache = new Map<string, Node>();
export function evaluate(expr: string, scope: Record<string, unknown>): unknown {
  let ast = cache.get(expr);
  if (!ast) { ast = new Parser(tokenize(expr)).parse(); cache.set(expr, ast); }
  return ev(ast, scope);
}
